plugin/fulltext/ngram_parser/plugin_ngram.cc

```diff
@@ -0,0 +1,238 @@
+/*
+   Copyright (c) 2014, 2017, Oracle and/or its affiliates. All rights reserved.
+   Copyright (c) 2020, MariaDB.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License, version 2.0,
+   as published by the Free Software Foundation.
+
+   This program is also distributed with certain software (including
+   but not limited to OpenSSL) that is licensed under separate terms,
+   as designated in a particular file or component or in included license
+   documentation.  The authors of MySQL hereby grant you an additional
+   permission to link the program and your derivative works with the
+   separately licensed software that they have included with MySQL.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License, version 2.0, for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+#include <stddef.h>
+#include <limits.h>
+#include <my_attribute.h>
+#include <my_global.h>
+#include "../storage/innobase/include/fts0tokenize.h"
+
+#define is_alnum(ct)    ((ct) & (_MY_U | _MY_L | _MY_NMR))
+
+static int ngram_token_size;
+
+static MYSQL_SYSVAR_INT(token_size, ngram_token_size,
+  PLUGIN_VAR_READONLY,
+  "Ngram full text plugin parser token size in characters",
+  nullptr, nullptr, 2, 1, 10, 0);
+
+// Splits a string into ngrams and emits them.
+static int split_into_ngrams(MYSQL_FTPARSER_PARAM *param, const uchar *doc,
+                             int len, MYSQL_FTPARSER_BOOLEAN_INFO *bool_info)
+{
+  const CHARSET_INFO *cs= param->cs;
+  const uchar *start= doc;
+  const uchar *end= doc + len;
+  const uchar *next= start;
+  int n_chars= 0;
+  int ret= 0;
+  bool is_first= true;
+
+  assert(cs->mbminlen == 1);
```

Comment:
I think that this should be tested with `ucs2`, `utf16`, `utf16le`, for which `mbminlen=2` holds.

That is, either this code should stop assuming that the encoding is a superset of ASCII, or we should reject incompatible encodings or collations earlier.

Reply:
Removed those assertions from the plugin code, since it's now should work with any character sets.
But currently server code itself prevents fulltext indexes from being created on columns with charsets with `mbminlen>=2`. I don't really know why.